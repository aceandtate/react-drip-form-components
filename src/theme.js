export const colors = {
  white: 'rgb(255, 255, 255)',
  grayDark: 'rgb(54, 54, 54)',
  gray: 'rgb(204, 204, 204)',
  grayLight: 'rgb(187, 187, 187)',
  grayLightest: 'rgb(219, 219, 219)',
  primary: 'rgb(0, 188, 212)',
  error: 'rgb(245, 0, 87)',
  disabled: 'rgb(245, 245, 245)',
  shadow: 'rgba(0, 0, 0, 0.04)',
};

export const theme = {
  // typography
  fontSize: '1rem',
  color: colors.grayDark,
  error: colors.error,
  disabled: colors.gray,

  // background
  background: colors.white,
  backgroundEnter: colors.white,
  backgroundFocus: colors.white,
  backgroundDisabled: colors.disabled,
  backgroundError: colors.white,

  // border
  borderWidth: '1px',
  borderRadius: '3px',
  borderColor: colors.grayLightest,
  borderColorEnter: colors.gray,
  borderColorFocus: colors.primary,
  borderColorDisabled: colors.disabled,
  borderColorError: colors.error,

  // box-shadow
  boxShadow: '0 0 0 transparent',
  boxShadowEnter: `0 2px 2px ${colors.shadow} inset`,
  boxShadowDisabled: `0 2px 2px ${colors.shadow} inset`,

  // transition
  transitionProperty: 'color, box-shadow, border',
  transitionDuration: '110ms',
  transitionTimingFunction: 'cubic-bezier(.25, .46, .45, .94)', // ease-out-quad

  // control
  controlHeight: '2.7em',

  // select
  selectArrowColor: colors.primary,

  // spinner
  spinnerColor: colors.primary,
  spinnerDuration: '500ms',
};

export const extendTheme = (extendsTheme = {}) => ({
  ...theme,
  ...extendsTheme,
});

export const themeProps = {
  theme: {
    rdf: theme,
  },
};

export const get = (props, key) => props.theme.rdf[key];

export const bind = key => props => get(props, key);
