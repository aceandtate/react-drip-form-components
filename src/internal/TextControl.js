import styled from 'styled-components';
import { themeProps, get, bind } from '../theme';

const TextControl = styled.input`
  box-sizing: border-box;
  display: block;
  width: 100%;
  height: ${bind('controlHeight')};
  appearance: none;
  margin: 0;
  padding: 0.3em 0.7em;
  background: ${bind('background')};
  border-style: solid;
  border-width: ${bind('borderWidth')};
  border-color: ${props => props.error ? get(props, 'borderColorError') : get(props, 'borderColor')};
  border-radius: ${bind('borderRadius')};
  box-shadow: ${bind('boxShadow')};
  color: ${bind('color')};
  vertical-align: top;
  font-size: 1rem;
  font-family: inherit;
  line-height: 1.4;
  transition-property: ${bind('transitionProperty')};
  transition-duration: ${bind('transitionDuration')};
  transition-timing-function: ${bind('transitionTimingFunction')};

  &:not(:disabled):hover {
    background: ${props => props.error ? get(props, 'backgroundError') : get(props, 'backgroundEnter')};
    border-color: ${props => props.error ? get(props, 'borderColorError') : get(props, 'borderColorEnter')};
  }

  &:not(:disabled):focus {
    outline: none;
    background: ${props => props.error ? get(props, 'backgroundError') : get(props, 'backgroundFocus')};
    border-color: ${props => props.error ? get(props, 'borderColorError') : get(props, 'borderColorFocus')};
    box-shadow: ${bind('boxShadowEnter')};
  }

  &:disabled {
    color: ${bind('disabled')};
    background: ${bind('backgroundDisabled')};
    border-color: ${bind('borderColorDisabled')};
    cursor: not-allowed;
  }
`;

TextControl.defaultProps = themeProps;

export default TextControl;
