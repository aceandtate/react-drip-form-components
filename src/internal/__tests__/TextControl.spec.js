import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import { theme } from '../../theme';
import TextControl from '../TextControl';
import { mountWithTheme } from '../../__tests__/utils';


describe('Internal#<TextControl />', () => {
  test('Should be render children', () => {
    const element = <div id="foo">foo bar baz</div>;
    const wrapper = shallow(
      <TextControl>
        {element}
      </TextControl>
    );

    expect(wrapper.contains(element)).toBe(true);
  });


  test('Should be styled component', () => {
    const wrapper = shallow(<TextControl />);
    expect(wrapper).toHaveStyleRule('height', theme.controlHeight);
    expect(wrapper).toHaveStyleRule('background', theme.background);
    expect(wrapper).toHaveStyleRule('border-width', theme.borderWidth);
    expect(wrapper).toHaveStyleRule('border-color', theme.borderColor);
    expect(wrapper).toHaveStyleRule('border-radius', theme.borderRadius);
    expect(wrapper).toHaveStyleRule('box-shadow', theme.boxShadow);
    expect(wrapper).toHaveStyleRule('color', theme.color);
    expect(wrapper).toHaveStyleRule('transition-property', theme.transitionProperty);
    expect(wrapper).toHaveStyleRule('transition-duration', theme.transitionDuration);
    expect(wrapper).toHaveStyleRule('transition-timing-function', theme.transitionTimingFunction);

    wrapper.setProps({ error: 'has error' });
    expect(wrapper).toHaveStyleRule('border-color', theme.borderColorError);
  });


  test('Should be changed styled component', () => {
    const testTheme = {
      controlHeight: '10rem',
      background: 'black',
      borderWidth: '1em',
      borderColor: 'pink',
      borderColorError: 'orangered',
      borderRadius: '20px',
      boxShadow: '10px 10px 100px black',
      color: 'ivory',
      transitionProperty: 'font-size, transform',
      transitionDuration: '100s',
      transitionTimingFunction: 'linear',
    };

    let wrapper;

    wrapper = mountWithTheme(<TextControl />, testTheme);
    expect(wrapper).toHaveStyleRule('height', testTheme.controlHeight);
    expect(wrapper).toHaveStyleRule('background', testTheme.background);
    expect(wrapper).toHaveStyleRule('border-width', testTheme.borderWidth);
    expect(wrapper).toHaveStyleRule('border-color', testTheme.borderColor);
    expect(wrapper).toHaveStyleRule('border-radius', testTheme.borderRadius);
    expect(wrapper).toHaveStyleRule('box-shadow', testTheme.boxShadow);
    expect(wrapper).toHaveStyleRule('color', testTheme.color);
    expect(wrapper).toHaveStyleRule('transition-property', testTheme.transitionProperty);
    expect(wrapper).toHaveStyleRule('transition-duration', testTheme.transitionDuration);
    expect(wrapper).toHaveStyleRule('transition-timing-function', testTheme.transitionTimingFunction);

    wrapper = mountWithTheme(<TextControl error="has error" />, testTheme);
    expect(wrapper).toHaveStyleRule('border-color', testTheme.borderColorError);
  });
});
