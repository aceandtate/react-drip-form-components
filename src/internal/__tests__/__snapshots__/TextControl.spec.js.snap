// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Internal#<TextControl /> Should be changed styled component 1`] = `
.c0 {
  box-sizing: border-box;
  display: block;
  width: 100%;
  height: 10rem;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0;
  padding: 0.3em 0.7em;
  background: black;
  border-style: solid;
  border-width: 1em;
  border-color: pink;
  border-radius: 20px;
  box-shadow: 10px 10px 100px black;
  color: ivory;
  vertical-align: top;
  font-size: 1rem;
  font-family: inherit;
  line-height: 1.4;
  -webkit-transition-property: font-size,-webkit-transform;
  -webkit-transition-property: font-size,transform;
  transition-property: font-size,transform;
  -webkit-transition-duration: 100s;
  transition-duration: 100s;
  -webkit-transition-timing-function: linear;
  transition-timing-function: linear;
}

.c0:not(:disabled):hover {
  background: rgb(255,255,255);
  border-color: rgb(204,204,204);
}

.c0:not(:disabled):focus {
  outline: none;
  background: rgb(255,255,255);
  border-color: rgb(0,188,212);
  box-shadow: 0 2px 2px rgba(0,0,0,0.04) inset;
}

.c0:disabled {
  color: rgb(204,204,204);
  background: rgb(245,245,245);
  border-color: rgb(245,245,245);
  cursor: not-allowed;
}

<ThemeProvider
  theme={
    Object {
      "rdf": Object {
        "background": "black",
        "backgroundDisabled": "rgb(245, 245, 245)",
        "backgroundEnter": "rgb(255, 255, 255)",
        "backgroundError": "rgb(255, 255, 255)",
        "backgroundFocus": "rgb(255, 255, 255)",
        "borderColor": "pink",
        "borderColorDisabled": "rgb(245, 245, 245)",
        "borderColorEnter": "rgb(204, 204, 204)",
        "borderColorError": "orangered",
        "borderColorFocus": "rgb(0, 188, 212)",
        "borderRadius": "20px",
        "borderWidth": "1em",
        "boxShadow": "10px 10px 100px black",
        "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "color": "ivory",
        "controlHeight": "10rem",
        "disabled": "rgb(204, 204, 204)",
        "error": "rgb(245, 0, 87)",
        "fontSize": "1rem",
        "selectArrowColor": "rgb(0, 188, 212)",
        "spinnerColor": "rgb(0, 188, 212)",
        "spinnerDuration": "500ms",
        "transitionDuration": "100s",
        "transitionProperty": "font-size, transform",
        "transitionTimingFunction": "linear",
      },
    }
  }
>
  <styled.input
    theme={
      Object {
        "rdf": Object {
          "background": "rgb(255, 255, 255)",
          "backgroundDisabled": "rgb(245, 245, 245)",
          "backgroundEnter": "rgb(255, 255, 255)",
          "backgroundError": "rgb(255, 255, 255)",
          "backgroundFocus": "rgb(255, 255, 255)",
          "borderColor": "rgb(219, 219, 219)",
          "borderColorDisabled": "rgb(245, 245, 245)",
          "borderColorEnter": "rgb(204, 204, 204)",
          "borderColorError": "rgb(245, 0, 87)",
          "borderColorFocus": "rgb(0, 188, 212)",
          "borderRadius": "3px",
          "borderWidth": "1px",
          "boxShadow": "0 0 0 transparent",
          "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "color": "rgb(54, 54, 54)",
          "controlHeight": "2.7em",
          "disabled": "rgb(204, 204, 204)",
          "error": "rgb(245, 0, 87)",
          "fontSize": "1rem",
          "selectArrowColor": "rgb(0, 188, 212)",
          "spinnerColor": "rgb(0, 188, 212)",
          "spinnerDuration": "500ms",
          "transitionDuration": "110ms",
          "transitionProperty": "color, box-shadow, border",
          "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
        },
      }
    }
  >
    <input
      className="c0"
    />
  </styled.input>
</ThemeProvider>
`;

exports[`Internal#<TextControl /> Should be changed styled component 2`] = `
.c0 {
  box-sizing: border-box;
  display: block;
  width: 100%;
  height: 10rem;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0;
  padding: 0.3em 0.7em;
  background: black;
  border-style: solid;
  border-width: 1em;
  border-color: orangered;
  border-radius: 20px;
  box-shadow: 10px 10px 100px black;
  color: ivory;
  vertical-align: top;
  font-size: 1rem;
  font-family: inherit;
  line-height: 1.4;
  -webkit-transition-property: font-size,-webkit-transform;
  -webkit-transition-property: font-size,transform;
  transition-property: font-size,transform;
  -webkit-transition-duration: 100s;
  transition-duration: 100s;
  -webkit-transition-timing-function: linear;
  transition-timing-function: linear;
}

.c0:not(:disabled):hover {
  background: rgb(255,255,255);
  border-color: orangered;
}

.c0:not(:disabled):focus {
  outline: none;
  background: rgb(255,255,255);
  border-color: orangered;
  box-shadow: 0 2px 2px rgba(0,0,0,0.04) inset;
}

.c0:disabled {
  color: rgb(204,204,204);
  background: rgb(245,245,245);
  border-color: rgb(245,245,245);
  cursor: not-allowed;
}

<ThemeProvider
  theme={
    Object {
      "rdf": Object {
        "background": "black",
        "backgroundDisabled": "rgb(245, 245, 245)",
        "backgroundEnter": "rgb(255, 255, 255)",
        "backgroundError": "rgb(255, 255, 255)",
        "backgroundFocus": "rgb(255, 255, 255)",
        "borderColor": "pink",
        "borderColorDisabled": "rgb(245, 245, 245)",
        "borderColorEnter": "rgb(204, 204, 204)",
        "borderColorError": "orangered",
        "borderColorFocus": "rgb(0, 188, 212)",
        "borderRadius": "20px",
        "borderWidth": "1em",
        "boxShadow": "10px 10px 100px black",
        "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "color": "ivory",
        "controlHeight": "10rem",
        "disabled": "rgb(204, 204, 204)",
        "error": "rgb(245, 0, 87)",
        "fontSize": "1rem",
        "selectArrowColor": "rgb(0, 188, 212)",
        "spinnerColor": "rgb(0, 188, 212)",
        "spinnerDuration": "500ms",
        "transitionDuration": "100s",
        "transitionProperty": "font-size, transform",
        "transitionTimingFunction": "linear",
      },
    }
  }
>
  <styled.input
    error="has error"
    theme={
      Object {
        "rdf": Object {
          "background": "rgb(255, 255, 255)",
          "backgroundDisabled": "rgb(245, 245, 245)",
          "backgroundEnter": "rgb(255, 255, 255)",
          "backgroundError": "rgb(255, 255, 255)",
          "backgroundFocus": "rgb(255, 255, 255)",
          "borderColor": "rgb(219, 219, 219)",
          "borderColorDisabled": "rgb(245, 245, 245)",
          "borderColorEnter": "rgb(204, 204, 204)",
          "borderColorError": "rgb(245, 0, 87)",
          "borderColorFocus": "rgb(0, 188, 212)",
          "borderRadius": "3px",
          "borderWidth": "1px",
          "boxShadow": "0 0 0 transparent",
          "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "color": "rgb(54, 54, 54)",
          "controlHeight": "2.7em",
          "disabled": "rgb(204, 204, 204)",
          "error": "rgb(245, 0, 87)",
          "fontSize": "1rem",
          "selectArrowColor": "rgb(0, 188, 212)",
          "spinnerColor": "rgb(0, 188, 212)",
          "spinnerDuration": "500ms",
          "transitionDuration": "110ms",
          "transitionProperty": "color, box-shadow, border",
          "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
        },
      }
    }
  >
    <input
      className="c0"
    />
  </styled.input>
</ThemeProvider>
`;

exports[`Internal#<TextControl /> Should be styled component 1`] = `
.c0 {
  box-sizing: border-box;
  display: block;
  width: 100%;
  height: 2.7em;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0;
  padding: 0.3em 0.7em;
  background: rgb(255,255,255);
  border-style: solid;
  border-width: 1px;
  border-color: rgb(219,219,219);
  border-radius: 3px;
  box-shadow: 0 0 0 transparent;
  color: rgb(54,54,54);
  vertical-align: top;
  font-size: 1rem;
  font-family: inherit;
  line-height: 1.4;
  -webkit-transition-property: color,box-shadow,border;
  transition-property: color,box-shadow,border;
  -webkit-transition-duration: 110ms;
  transition-duration: 110ms;
  -webkit-transition-timing-function: cubic-bezier(.25,.46,.45,.94);
  transition-timing-function: cubic-bezier(.25,.46,.45,.94);
}

.c0:not(:disabled):hover {
  background: rgb(255,255,255);
  border-color: rgb(204,204,204);
}

.c0:not(:disabled):focus {
  outline: none;
  background: rgb(255,255,255);
  border-color: rgb(0,188,212);
  box-shadow: 0 2px 2px rgba(0,0,0,0.04) inset;
}

.c0:disabled {
  color: rgb(204,204,204);
  background: rgb(245,245,245);
  border-color: rgb(245,245,245);
  cursor: not-allowed;
}

<input
  className="c0"
/>
`;

exports[`Internal#<TextControl /> Should be styled component 2`] = `
.c0 {
  box-sizing: border-box;
  display: block;
  width: 100%;
  height: 2.7em;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0;
  padding: 0.3em 0.7em;
  background: rgb(255,255,255);
  border-style: solid;
  border-width: 1px;
  border-color: rgb(245,0,87);
  border-radius: 3px;
  box-shadow: 0 0 0 transparent;
  color: rgb(54,54,54);
  vertical-align: top;
  font-size: 1rem;
  font-family: inherit;
  line-height: 1.4;
  -webkit-transition-property: color,box-shadow,border;
  transition-property: color,box-shadow,border;
  -webkit-transition-duration: 110ms;
  transition-duration: 110ms;
  -webkit-transition-timing-function: cubic-bezier(.25,.46,.45,.94);
  transition-timing-function: cubic-bezier(.25,.46,.45,.94);
}

.c0:not(:disabled):hover {
  background: rgb(255,255,255);
  border-color: rgb(245,0,87);
}

.c0:not(:disabled):focus {
  outline: none;
  background: rgb(255,255,255);
  border-color: rgb(245,0,87);
  box-shadow: 0 2px 2px rgba(0,0,0,0.04) inset;
}

.c0:disabled {
  color: rgb(204,204,204);
  background: rgb(245,245,245);
  border-color: rgb(245,245,245);
  cursor: not-allowed;
}

<input
  className="c0"
/>
`;
